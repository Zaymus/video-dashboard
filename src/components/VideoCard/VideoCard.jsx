import styled from "styled-components";
import Thumbnail from '../Thumbnail';
import Loader from "../common/Loader";
import { Suspense } from "react";

const Card = styled.div`
  display: flex;
  flex-direction: column;
  width: calc(var(--videos-usable-width) / var(--videos-per-row) - var(--videos-gutter-margin) - 0.1px);
  cursor: pointer;
  border-radius: 15px;
  padding: 20px;

  &:hover {
    background-color: var(--border-dark);
  }
`;

const Title = styled.p`
  margin: 0;
  font-size: 16px;
  line-height: 22px;
  white-space: pre-wrap;
  margin-top: 5px;
`;

const VideoCard = ({ id, title, thumbnails }) => {
  return (
    <Card key={id}>
      <Suspense fallback={<Loader />}>
        <Thumbnail thumbnails={thumbnails} />
      </Suspense>
      <Title>{title}</Title>
    </Card>
  );
}

export default VideoCard;