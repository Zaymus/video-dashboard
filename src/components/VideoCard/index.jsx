import { lazy } from 'react';

const VideoCard = lazy(() => import('./VideoCard'));

export default VideoCard;