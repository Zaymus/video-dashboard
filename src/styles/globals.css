:root {
  --primary: #1E90FF;
  --secondary: #FF6B6B;
  --background-dark: #121212;
  --surface-dark: #1F1F1F;
  --text-primary-dark: #FFFFFF;
  --text-secondary-dark: #B3B3B3;
  --border-dark: #2C2C2C;
  --success: #2ECC71;
  --error: #E74C3C;
  --warning: #F1C40F;
  --background-light: #F9F9F9;
  --surface-light: #FFFFFF;
  --text-primary-light: #1A1A1A;
  --text-secondary-light: #4F4F4F;
  --border-light: #E0E0E0;

  --videos-gutter-margin: 30px;
  --videos-per-row: 5;
  --videos-usable-width: calc(100% - var(--videos-gutter-margin) * 2);
}

body {
  background-color: var(--background-dark);
  margin: 0;
  color: var(--text-primary-dark);
  font-size: 16pt;
}

#root {
  min-height: 100vh;
}