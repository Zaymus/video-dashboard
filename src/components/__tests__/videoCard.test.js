import { jest, describe, test, expect, beforeEach } from "@jest/globals";
import { render, act, fireEvent, screen, waitFor } from "@testing-library/react";
import '@testing-library/jest-dom';

const mockNavigate = jest.fn();
jest.mock('react-router', () => ({
  ...jest.requireActual('react-router'),
  useNavigate: () => mockNavigate,
}));

import VideoCard from "../VideoCard/VideoCard.jsx";

const mockVideo = {
  id: '12345',
  title: 'Test Video',
  thumbnails: {
    standard: { url: 'https://cdn.theatlantic.com/media/mt/science/cat_caviar.jpg' },
  },
}

describe('VideoCard Component', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  test('matches snapshot', () => {
    const { container } = render(<VideoCard id={mockVideo.id} title={mockVideo.title} thumbnails={mockVideo.thumbnails} />);
    expect(container).toMatchSnapshot();
  });

  test('clickHandler works as expected', () => {
    const { container } = render(<VideoCard id={mockVideo.id} title={mockVideo.title} thumbnails={mockVideo.thumbnails} />);
    
    act(() => {
      fireEvent.click(container.querySelector('div'));
    });

    expect(mockNavigate).toHaveBeenCalledWith(`/video/${mockVideo.id}`);
  });

  describe('Thumbnail rendering', () => {
    test('standard', async () => {
      render(<VideoCard id={mockVideo.id} title={mockVideo.title} thumbnails={mockVideo.thumbnails} />);

      await waitFor(() => {
        expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
      });

      var thumbnail = screen.getByAltText('Thumbnail image');
      expect(thumbnail).toBeInTheDocument();
      expect(thumbnail.src).toBe(mockVideo.thumbnails.standard.url);
    });

    test('medium', async () => {
      mockVideo.thumbnails.medium = { url: 'https://img3.stockfresh.com/files/n/nyul/m/16/623420_stock-photo-portrait-of-happy-old-man.jpg' };
      render(<VideoCard id={mockVideo.id} title={mockVideo.title} thumbnails={mockVideo.thumbnails} />);

      await waitFor(() => {
        expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
      });

      var thumbnail = screen.getByAltText('Thumbnail image');
      expect(thumbnail).toBeInTheDocument();
      expect(thumbnail.src).toBe(mockVideo.thumbnails.medium.url);
    });

    test('high', async () => {
      mockVideo.thumbnails.high = { url: 'https://previews.123rf.com/images/bowie15/bowie151208/bowie15120800108/15112437-businessman-running.jpg' };
      render(<VideoCard id={mockVideo.id} title={mockVideo.title} thumbnails={mockVideo.thumbnails} />);

      await waitFor(() => {
        expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
      });

      var thumbnail = screen.getByAltText('Thumbnail image');
      expect(thumbnail).toBeInTheDocument();
      expect(thumbnail.src).toBe(mockVideo.thumbnails.high.url);
    });

    test('maxres', async () => {
      mockVideo.thumbnails.maxres = { url: 'https://ichef.bbci.co.uk/ace/standard/976/cpsprodpb/514F/production/_95051802_olddriver.jpg' };
      render(<VideoCard id={mockVideo.id} title={mockVideo.title} thumbnails={mockVideo.thumbnails} />);

      await waitFor(() => {
        expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
      });

      var thumbnail = screen.getByAltText('Thumbnail image');
      expect(thumbnail).toBeInTheDocument();
      expect(thumbnail.src).toBe(mockVideo.thumbnails.maxres.url);
    });
  });
});