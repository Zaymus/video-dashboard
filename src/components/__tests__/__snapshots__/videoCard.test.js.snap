// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`VideoCard Component matches snapshot 1`] = `
<div>
  <div
    class="sc-fhHczv kHoESd"
  >
    <div
      class="sc-bRKDuR fMrgJQ"
      role="progressbar"
    >
      <svg
        class="spinner"
        viewBox="0 0 50 50"
      >
        <circle
          class="path"
          cx="25"
          cy="25"
          fill="none"
          r="20"
          stroke-width="5"
        />
      </svg>
    </div>
    <p
      class="sc-hvigdm dToLmu"
    >
      Test Video
    </p>
  </div>
</div>
`;
